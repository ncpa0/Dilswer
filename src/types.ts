import type { BasicDataTypes, dataTypeSymbol } from "./type-def-helpers";
import type { ValueOf } from "./type-utils";

export type ArrayOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly arrayOf: DT;
};

export type RecordOf<TS extends RecordTypeSchema = RecordTypeSchema> = {
  readonly [dataTypeSymbol]: true;
  readonly recordOf: TS;
};

export type SetOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly setOf: DT;
};

export type OneOf<DT extends AllDataTypes[] = any[]> = {
  readonly [dataTypeSymbol]: true;
  readonly oneOf: DT;
};

export type Literal<
  V extends string | number | boolean = string | number | boolean
> = {
  readonly [dataTypeSymbol]: true;
  readonly literal: V;
};

export type Enum<E = any> = {
  readonly [dataTypeSymbol]: true;
  readonly enumInstance: E;
};

export type EnumMember<M = any> = {
  readonly [dataTypeSymbol]: true;
  readonly enumMember: M;
};

export type BasicDataType = ValueOf<typeof BasicDataTypes>;

export type ComplexDataType =
  | ArrayOf
  | RecordOf
  | SetOf
  | OneOf
  | Literal
  | Enum
  | EnumMember;

export type AllDataTypes = BasicDataType | ComplexDataType;

export type FieldDescriptor = {
  readonly required?: boolean;
  readonly type: AllDataTypes;
};

export interface RecordTypeSchema {
  readonly [key: string]: FieldDescriptor | AllDataTypes;
}
