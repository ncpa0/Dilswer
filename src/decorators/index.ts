export { ClassWithValidation, validate } from "./validation-decorator";
