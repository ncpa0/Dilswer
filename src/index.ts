import { DataType, getMetadata } from "@DataTypes/data-types";
import {
  And,
  Exclude,
  Omit,
  OptionalField,
  Partial,
  Pick,
  Required,
} from "@Intrinsic";
import { toJsonSchema } from "@JSONSchemaParser/to-json-schema";
import { toTsType } from "@TsTypeGenerator/to-ts-type";
import {
  createTypeGuardedFunction,
  createValidatedFunction,
} from "@Validation/create-validated-function";
import { createChecker, createValidator } from "@Validation/create-validator";
import { ensureDataType } from "@Validation/ensure-data-type";
export type { GetDataType } from "@DataTypes/type-utils";
export type {
  AllDataTypes,
  AllOf,
  AnyDataType,
  ArrayOf,
  BasicDataType,
  BasicTypeNames,
  ComplexDataType,
  Custom,
  Dict,
  Enum,
  EnumMember,
  FieldDescriptor,
  Literal,
  OneOf,
  RecordOf,
  RecordTypeSchema,
  SetOf,
  TypeMetadata,
} from "@DataTypes/types";
export * from "@Intrinsic";
export type { ParseToJsonSchemaOptions } from "@JSONSchemaParser/to-json-schema";
export type {
  TsParsingMode,
  TsParsingOptions,
} from "@TsTypeGenerator/parsers/parse";
export {
  And,
  assertDataType,
  createChecker,
  createTypeGuardedFunction,
  createValidatedFunction,
  createValidator,
  DataType,
  ensureDataType,
  Exclude,
  getMetadata,
  Omit,
  OptionalField,
  Partial,
  Pick,
  Required,
  toJsonSchema,
  toTsType,
};

const assertDataType = ensureDataType;

export default {
  And,
  assertDataType,
  createChecker,
  createTypeGuardedFunction,
  createValidatedFunction,
  createValidator,
  DataType,
  ensureDataType,
  Exclude,
  getMetadata,
  Omit,
  OptionalField,
  Partial,
  Pick,
  Required,
  toJsonSchema,
  toTsType,
};
