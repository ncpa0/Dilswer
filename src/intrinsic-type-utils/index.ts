export * from "@Intrinsic/and";
export * from "@Intrinsic/exclude";
export * from "@Intrinsic/omit";
export * from "@Intrinsic/partial";
export * from "@Intrinsic/pick";
export * from "@Intrinsic/required";
