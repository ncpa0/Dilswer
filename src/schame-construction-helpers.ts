import { AllDataTypes } from ".";
import type {
  ArrayOf,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  SetOf,
  TypeSchema,
} from "./types";

export const BasicDataTypes = {
  Unknown: "unknown",
  String: "string",
  Number: "number",
  Boolean: "boolean",
  Symbol: "symbol",
  Function: "function",
  Null: "null",
  Undefined: "undefined",
} as const;

export const DataType = {
  ...BasicDataTypes,
  RecordOf<TS extends TypeSchema>(args: TS): RecordOf<TS> {
    return { recordOf: args };
  },
  ArrayOf<DT extends AllDataTypes[]>(...args: DT): ArrayOf<DT> {
    return { arrayOf: args };
  },
  SetOf<DT extends AllDataTypes[]>(...args: DT): SetOf<DT> {
    return { setOf: args };
  },
  OneOf<DT extends AllDataTypes[]>(...args: DT): OneOf<DT> {
    return { oneOf: args };
  },
  Literal<V extends string | number | boolean>(value: V): Literal<V> {
    return { literal: value };
  },
  EnumMember<M extends number | string>(enumMember: M): EnumMember<M> {
    return { enumMember };
  },
  Enum<T extends string, TEnumValue extends string | number>(enumInstance: {
    [key in T]: TEnumValue;
  }): Enum<TEnumValue> {
    // @ts-expect-error
    return { enumInstance };
  },
};
