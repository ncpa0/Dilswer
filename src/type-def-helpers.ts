import type { AllDataTypes } from ".";
import type {
  ArrayOf,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  RecordTypeSchema,
  SetOf,
} from "./types";

export const dataTypeSymbol: unique symbol = Symbol();

export const BasicDataTypes = {
  Unknown: "unknown",
  String: "string",
  Number: "number",
  Boolean: "boolean",
  Symbol: "symbol",
  Function: "function",
  Null: "null",
  Undefined: "undefined",
} as const;

export const DataType = {
  ...BasicDataTypes,
  RecordOf<TS extends RecordTypeSchema>(args: TS): RecordOf<TS> {
    return {
      [dataTypeSymbol]: true,
      recordOf: args,
    };
  },
  ArrayOf<DT extends AllDataTypes[]>(...args: DT): ArrayOf<DT> {
    return {
      [dataTypeSymbol]: true,
      arrayOf: args,
    };
  },
  SetOf<DT extends AllDataTypes[]>(...args: DT): SetOf<DT> {
    return {
      [dataTypeSymbol]: true,
      setOf: args,
    };
  },
  OneOf<DT extends AllDataTypes[]>(...args: DT): OneOf<DT> {
    return {
      [dataTypeSymbol]: true,
      oneOf: args,
    };
  },
  Literal<V extends string | number | boolean>(value: V): Literal<V> {
    return {
      [dataTypeSymbol]: true,
      literal: value,
    };
  },
  EnumMember<M extends number | string>(enumMember: M): EnumMember<M> {
    return {
      [dataTypeSymbol]: true,
      enumMember,
    };
  },
  Enum<T extends string, TEnumValue extends string | number>(enumInstance: {
    [key in T]: TEnumValue;
  }): Enum<TEnumValue> {
    // @ts-expect-error
    return { enumInstance };
  },
};
