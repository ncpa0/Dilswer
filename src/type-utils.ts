import { AllDataTypes } from ".";
import type {
  ArrayOf,
  BasicDataType,
  ComplexDataType,
  Enum,
  EnumMember,
  Literal,
  OneOf,
  RecordOf,
  SetOf,
  TypeSchema,
} from "./types";

export type ValueOf<R extends Record<any, any>> = R extends Record<any, infer T>
  ? T
  : never;

export type UnknownFunction = (...args: unknown[]) => unknown;

export type ReWrap<T> = T extends Function
  ? T
  : T extends Set<infer ST>
  ? Set<ReWrap<ST>>
  : T extends object
  ? T extends infer O
    ? { [K in keyof O]: ReWrap<O[K]> }
    : never
  : T;

export type EnsureStringType<T> = T extends string ? T : string;

export type ExcludeRequired<S extends TypeSchema> = EnsureStringType<
  Exclude<
    ValueOf<{
      [K in keyof S]: S[K]["required"] extends false ? K : undefined;
    }>,
    undefined
  >
>;

export type ExcludeOptional<S extends TypeSchema> = EnsureStringType<
  Exclude<
    ValueOf<{
      [K in keyof S]: S[K]["required"] extends false ? undefined : K;
    }>,
    undefined
  >
>;

export type EnsureIsKey<K> = K extends
  | "arrayOf"
  | "recordOf"
  | "setOf"
  | "oneOf"
  | "literal"
  | "enumInstance"
  | "enumMember"
  ? K
  : "recordOf";

export type GetTypeFromArrayOf<D extends ComplexDataType> = D extends ArrayOf<
  infer T
>
  ? T[number]
  : never;

export type GetFieldDescriptorsFromSetOf<D extends ComplexDataType> =
  D extends SetOf<infer T> ? T[number] : never;

export type GetTypeFromRecordOf<D extends ComplexDataType> =
  D extends RecordOf<any> ? D : never;

export type GetTypeFromOneOf<D extends ComplexDataType> = D extends OneOf<
  infer T
>
  ? T[number]
  : never;

export type GetTypeFromLiteral<D extends ComplexDataType> = D extends Literal<
  infer T
>
  ? T
  : never;

export type GetTypeFromEnum<D extends ComplexDataType> = D extends Enum<infer T>
  ? T
  : never;

export type GetTypeFromEnumMember<D extends ComplexDataType> =
  D extends EnumMember<infer T> ? T : never;

export type ParseComplexType<D extends ComplexDataType> = {
  arrayOf: Array<ParseDataType<GetTypeFromArrayOf<D>>>;
  recordOf: ParseRecordType<GetTypeFromRecordOf<D>>;
  setOf: Set<ParseDataType<GetFieldDescriptorsFromSetOf<D>>>;
  oneOf: ParseDataType<GetTypeFromOneOf<D>>;
  literal: GetTypeFromLiteral<D>;
  enumInstance: GetTypeFromEnum<D>;
  enumMember: GetTypeFromEnumMember<D>;
}[EnsureIsKey<keyof D>];

export type ParseBasicDataType<D extends BasicDataType> = {
  unknown: unknown;
  string: string;
  number: number;
  boolean: boolean;
  symbol: symbol;
  function: UnknownFunction;
  null: null;
  undefined: undefined;
}[D];

export type ParseDataType<D> = D extends BasicDataType
  ? ParseBasicDataType<D>
  : D extends ComplexDataType
  ? ParseComplexType<D>
  : never;

export type ParseRecordType<S extends RecordOf> = {
  [K in ExcludeRequired<S["recordOf"]>]?: ParseDataType<
    S["recordOf"][K]["type"]
  >;
} & {
  [K in ExcludeOptional<S["recordOf"]>]: ParseDataType<
    S["recordOf"][K]["type"]
  >;
};

export type GetDataType<D extends AllDataTypes> = ReWrap<ParseDataType<D>>;
