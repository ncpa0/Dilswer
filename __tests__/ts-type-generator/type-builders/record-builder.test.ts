import { DataType, OptionalField, toTsType } from "../../../src/index";

describe("TsRecordBuilder", () => {
  it("should correctly generate a simple record type", () => {
    const dt = DataType.RecordOf({
      foo: DataType.String,
      bar: OptionalField(DataType.Number),
    });

    const tsType = toTsType(dt);

    expect(tsType).toMatchSnapshot();
    expect(tsType).toBe(
      "export type RecordType = {\n  foo: string;\n  bar?: number;\n};\n",
    );
  });

  it("should correctly generate a simple record type with title and description", () => {
    const dt = DataType.RecordOf({
      foo: DataType.String,
      bar: OptionalField(DataType.Number),
    });

    dt.setTitle("MyRec");
    dt.setDescription("My record's description");

    dt.recordOf.foo.setDescription("The foo field");
    dt.recordOf.bar.type.setDescription("The bar field");

    const tsType = toTsType(dt);

    expect(tsType).toMatchSnapshot();
    expect(tsType).toBe(
      "/**\n * My record's description\n */\nexport type MyRec = {\n  /**\n   * The foo field\n   */\n  foo: string;\n  /**\n   * The bar field\n   */\n  bar?: number;\n};\n",
    );
  });
});
