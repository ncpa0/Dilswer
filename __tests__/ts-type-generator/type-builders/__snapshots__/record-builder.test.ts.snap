// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TsRecordBuilder should correctly generate a simple record type 1`] = `
"export type Record1 = {
  foo: string;
  bar?: number;
};
"
`;

exports[`TsRecordBuilder should correctly generate a simple record type with title and description 1`] = `
"/**
 * My record's description
 */
export type MyRec = {
  /**
   * The foo field
   */
  foo: string;
  /**
   * The bar field
   */
  bar?: number;
};
"
`;
